machine.py

```diff
@@ -0,0 +1,102 @@
+# SPDX-License-Identifier: LGPL-2.1+
+import argparse
+import pexpect
+from typing import List, Optional
+import sys
+import time
+import subprocess
+import signal
+import contextlib
+
+import os
+import mkosi
+from mkosi.backend import Distribution, MkosiArgs, MkosiException, ARG_DEBUG, tmp_dir
+
+class Machine():
+    def __init__(self):
+        # Remains None until image is built and booted
+        self.serial = None
+        self.args = None
+
+    def __enter__(self):
+        parsed_args = mkosi.parse_args(sys.argv[1:])
+        parsed_args = parsed_args["default"]
+        self.args = mkosi.load_args(parsed_args)
+
+        mkosi.prepend_to_environ_path(self.args.extra_search_paths)
+
+        if self.args.verb == "genkey":
+            mkosi.generate_secure_boot_key(self.args)
+
+        if self.args.verb == "bump":
+            mkosi.bump_image_version(self.args)
+
+        if self.args.verb in mkosi.MKOSI_COMMANDS_SUDO:
+            mkosi.check_root()
+            mkosi.unlink_output(self.args)
+
+        if self.args.verb == "build":
+            mkosi.check_output(self.args)
+
+        #if args.verb == "summary":
+            #print_summary(args)
+
+        if mkosi.needs_build(self.args):
+            mkosi.check_root()
+            mkosi.check_native(self.args)
+            mkosi.init_namespace(self.args)
+            manifest = mkosi.build_stuff(self.args)
+
+            if self.args.auto_bump:
+                mkosi.bump_image_version(self.args)
+
+            mkosi.save_manifest(self.args, manifest)
+
+            mkosi.print_output_size(self.args)
+
+        if self.args.verb in ("shell", "boot"):
+            self.serial = pexpect.spawnu(" ".join(mkosi.pre_run_shell(self.args)), logfile=sys.stdout, timeout=240)
+
+        if self.args.verb == "qemu":
+            with mkosi.pre_run_qemu(self.args) as cmdline:
```

Comment:
Hmm, so one thing I didn't consider here is that pexpect will continue immediately after starting the process which means we'll exit the context manager and the temporary file allocated by pre_run_qemu() will be removed.

The solution here is store a context stack as part of the Machine class that will keep the context manager alive for as long as needed. I'll explain this in more detail in the chat we'll have today.